toolchain/common/enum_base.h

```diff
@@ -0,0 +1,179 @@
+  }
+
+ protected:
+  // The default constructor is explicitly defaulted (and constexpr) as a
+  // protected constructor to allow derived classes to be constructed but not
+  // the base itself. This should only be used in the `Create` function below.
+  constexpr EnumBase() = default;
+
+  // Create an instance from the raw enumerator. This should only be used to
+  // initialize the named constants in the derived type.
+  static constexpr auto Create(RawEnumType value) -> EnumType {
+    EnumType result;
+    result.value_ = value;
+    return result;
+  }
+
+  // Convert to the underlying integer type. Derived types can choose to expose
+  // this as part of their API.
+  constexpr auto AsInt() const -> UnderlyingType {
+    return static_cast<UnderlyingType>(value_);
+  }
+
+  // Convert from the underlying integer type. Derived types can choose to
+  // expose this as part of their API.
+  static constexpr auto FromInt(UnderlyingType value) -> EnumType {
+    return Create(static_cast<RawEnumType>(value));
+  }
+
+ private:
+  static llvm::StringLiteral names[];
+
+  RawEnumType value_;
+};
+
+}  // namespace Carbon::Internal
+
+// Use this before defining a class that derives from `EnumBase` to begin the
+// definition of the raw `enum class`. It should be followed by the body of that
+// raw enum class.
+#define CARBON_DEFINE_RAW_ENUM_CLASS(EnumClassName, UnderlyingType) \
+  namespace Internal {                                              \
+  /* NOLINTNEXTLINE(bugprone-macro-parentheses) */                  \
+  enum class EnumClassName##RawEnum : UnderlyingType;               \
+  }                                                                 \
+  enum class ::Carbon::Internal::EnumClassName##RawEnum : UnderlyingType
+
+// In CARBON_DEFINE_RAW_ENUM_CLASS block, use this to generate each enumerator.
+#define CARBON_RAW_ENUM_ENUMERATOR(Name) Name,
+
+// Use this to compute the `Internal::EnumBase` specialization for a Carbon enum
+// class. It both computes the name of the raw enum and ensures all the
+// namespaces are correct.
+#define CARBON_ENUM_BASE(EnumClassName)       \
+  ::Carbon::Internal::EnumBase<EnumClassName, \
+                               ::Carbon::Internal::EnumClassName##RawEnum>
+
+// Use this within the Carbon enum class body to generate named constant
+// declarations for each value.
+#define CARBON_ENUM_CONSTANT_DECLARATION(Name) static const EnumType Name;
+
+// Use this immediately after the Carbon enum class body to define each named
+// constant.
+#define CARBON_ENUM_CONSTANT_DEFINITION(EnumClassName, Name) \
+  constexpr EnumClassName EnumClassName::Name =              \
+      EnumClassName::Create(RawEnumType::Name);
+
+// Use this in the `.cpp` file for an enum class to start the definition of the
+// constant names array for each enumerator. It is followed by the desired
+// constant initializer.
+#define CARBON_DEFINE_ENUM_CLASS_NAMES(EnumClassName)                         \
```

Comment:
```suggestion
//
// `clang-format` has a bug with spacing around `->` returns in macros. See
// https://bugs.llvm.org/show_bug.cgi?id=48320 for details.
#define CARBON_DEFINE_ENUM_CLASS_NAMES(EnumClassName)                         \
```

Suggesting to keep this comment since it comes up on line 166, and I feel it keeps coming up in reviews.

Reply:
Sure. I had hoped it would stop coming up in reviews as it will no longer show up in new code (this should be the only place where it happens).